### PR TITLE
CON-1072 : Add support for NodeGetVolumeStats

### PR DESCRIPTION
**Pull request by ckallianpur**
Signed-off-by: chaitra <chaitra.kallianpur@hpe.com>
Implement NodeGetVolumeStats in csi-driver and get the metrics from CSP

### CODE REVIEW

File: pkg/flavor/kubernetes/flavor.go [hunk 1]
```diff
@@ -36,9 +36,11 @@ import (
 )
 
 const (
-	allowOverrides       = "allowOverrides"
-	nodeUnreachableTaint = "node.kubernetes.io/unreachable"
-	nodeLostCondition    = "NodeLost"
+	allowOverrides                = "allowOverrides"
+	nodeUnreachableTaint          = "node.kubernetes.io/unreachable"
+	nodeLostCondition             = "NodeLost"
+	provisionerSecretNameKey      = "csi.storage.k8s.io/provisioner-secret-name"
```
Comment (raunakkumar) [inline pkg/flavor/kubernetes/flavor.go:42]: nit: these could move to constants.go

File: pkg/driver/node_server.go [hunk 2]
```diff
@@ -1654,7 +1656,62 @@ func (driver *Driver) NodeGetVolumeStats(ctx context.Context, in *csi.NodeGetVol
...
+			return nil, status.Errorf(codes.InvalidArgument, "volume path %s does not exist", volumePath)
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if notMount {
+		return nil, status.Errorf(codes.InvalidArgument, "volume path %s is not mounted", volumePath)
+	}
+
+	folders := strings.Split(in.VolumePath, "/")
+	pvName := folders[len(folders)-2]
```
Comment (raunakkumar) [inline pkg/driver/node_server.go:1678]: nit : validate the size of folders before accessing by index

**Review by raunakkumar — APPROVED** [review]
lgtm

File: pkg/driver/node_server.go [hunk 3]
```diff
@@ -1654,7 +1656,62 @@ func (driver *Driver) NodeGetVolumeStats(ctx context.Context, in *csi.NodeGetVol
...
+	return &csi.NodeGetVolumeStatsResponse{
+		Usage: []*csi.VolumeUsage{
+			{
+				Available: volume.FreeBytes,
+				Total:     volume.Size,
+				Used:      volume.UsedBytes,
+				Unit:      csi.VolumeUsage_BYTES,
+			},
+		},
+	}, nil
```
Comment (rgcostea) [inline pkg/driver/node_server.go:1714]: Please add inode usage as well.

Comment (ckallianpur) [reply pkg/driver/node_server.go:1714]: Currently  I am getting these fields from csp
 "id" : "064763f97c7c986c6600000000000000000000000f",
  "name" : "pvc-ae6b8013-61cb-4794-844a-468eb89b5ca6",
  "size" : 10737418240,
  "description" : "Volume from csi",
  "published" : true,
  "config" : {
    "pool" : "default",
    "folder" : "",
    "encrypted" : false,
    "thick" : false,
    "performance_policy" : "SQL Server",
    "dedupe_enabled" : false,
    "limit_iops" : 76800,
    "limit_mbps" : -1,
    "destroy_on_delete" : false,
    "sync_on_detach" : false,
    "clone_of" : "",
    "create_snapshot" : false,
    "target_scope" : "volume"
  },
  "snaphotUsedBytes" : 106496,
  "pool_name" : "default",
  "base_snapshot_id" : "",
  "volume_group_id" : "",
  "used_bytes" : 0,
  "free_bytes" : 10737418240,
  "reservation_ratio" : 0.0,
  "data_compression_saved_bytes" : 18567860,
  "snapshot_used_bytes" : 106496
}

Do I need to get inode get from array or is it one of the values already pulled?

Comment (rgcostea) [reply pkg/driver/node_server.go:1714]: Those don't come from the array.  See here: https://github.com/hpe-storage/csi-driver/pull/190/files

Comment (ckallianpur) [reply pkg/driver/node_server.go]: Made the changes.

File: pkg/driver/node_server.go (hunk 2 above)
Comment (ckallianpur) [reply pkg/driver/node_server.go:1683]: done

**Review by rgcostea — APPROVED** [review]

**Pull request merged**